docs/design/operator.md

```diff
@@ -0,0 +1,407 @@
+APIs defined as a set of Custom Resource Definitions (CRDs).
+
+## System Overview
+
+The Ceph-CSI Operator is a Kubernetes operator that offers a Kubernetes native
+management interfaces for Ceph CSI drivers. It provides a set of new Kubernetes
+APIs (CRDs) for governing deployment and management of Ceph-CSI drivers, as
+well as connection details to Ceph clusters. The operator ensures that the
+desired state specified in the CRDs is maintained in the cluster.
+
+## Assumptions and Dependencies
+
+Kubernetes cluster is available and operational. Ceph cluster is set up and
+accessible from the Kubernetes cluster. Users are familiar with Kubernetes and
+Ceph concepts. CRD and operator versions are compatible with the Kubernetes
+version in use.
+
+## General Constraints
+
+Resource limitations of the Kubernetes cluster (CPU, memory). Network latency
+and connectivity between the Kubernetes cluster and Ceph cluster. Watching for
+changes on the ceph cluster and updating internal details.  Security policies
+and RBAC configurations in the Kubernetes cluster.
+
+## Goals and Guidelines
+
+- Automate the deployment and lifecycle for Ceph-CSI drivers.
+- Provide a consistent and flexible configuration mechanism using CRDs.
+- Enable namespace-scoped configurations for different Ceph-CSI drivers.
+- Ensure high availability and scalability of the operator.
+- Support multiple deployment methods for the operator (single-namespace,
+  multi-namespace, all-namespaces).
+- Support multiple operator deployment in a single cluster when running in a
+  single-namespace mode.
+
+## Development Methods
+
+Use operator framework which is higly used in operator developed and also by
+following Kubernetes best practices for operator development.
+
+## Architecture
+
+```mermaid
+graph TD
+    A[CRD Changes] --> B[Operator]
+    B -->  D[Configure CephFS CSI Driver]
+    B -->  E[Configure RBD CSI Driver]
+    B -->  F[Configure NFS CSI Driver]
+
+```
+
+In this updated diagram:
+
+- **CRD Changes**: Represents changes made to Custom Resource Definitions,
+  which trigger actions in the operator.
+- **Operator**: Listens for changes in CRDs and initiates the installation and
+  configuration of CSI drivers.
+- **Configure CephFS, NFS, RBD**: Actions performed by the operator to install
+  and configure the respective CSI drivers based on the CRD changes.
+
+## CRD's for ceph-csi-operator
+
+### CephCSIOperatorConfig CRD
+
+Manages operator-level configurations and offers a place to overwrite settings
+for CSI drivers. This CRD is a namespace-scoped CRD and a single CR named
+ceph-csi-operator-config can be created by the admin in the namespace
+where operator is running. The configurations are catagorized into 2 different
+parts
+
+- Operator configurations
+- Common configurations across all CSI driver
+
+```yaml
+---
+kind: CephCSIOperatorConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: csioperatorconfig
+  namespace: operator-namespace
```

Comment:
How about using the namespace `ceph-csi` in the examples?
```suggestion
  namespace: ceph-csi
```

Reply:
i used operator-namespace as example to provide more clarity that we can have the CR's created in the operator-namespace, let me know if you still suggest to change it to ceph-csi.

Reply:
@travisn i choose to keep `operator-namespace` in the design doc which gives details on what was the design consideration but as you suggested with names that is something we will add it to the deployment section (in documentation) or the deployment yamls